Convert snapshot_bot to use ``ImageDatabase``s

alexdewar opened the pull request:
Closes #274.

The main reason for this change is that @efkag has been using ``snapshot_bot`` to record image databases and it would be convenient if any future datasets were in the standard ``ImageDatabase`` format to make it easier to analyse with our standard tools (e.g. navbench). An added bonus, though, is that it should also make analysing any navigation-related issues easier (e.g. by using a quiver plot from navbench).

neworderofjamie commented on projects/snapshot_bot/memory.cc:
```diff
@@ -1,26 +1,31 @@
 using namespace units::literals;
 using namespace units::math;
 
+constexpr const char *CSVBestHeading = "Best heading [degrees]";
```
I'm not 100% on the rules for ``constexpr`` objects but I think these should probably be in the anonymous namespace so they're only module scope

neworderofjamie commented on projects/snapshot_bot/memory.h:
```diff
@@ -10,19 +10,31 @@
 
+// Bounds used for extracting masks from ODK2 images
+const cv::Scalar odk2MaskLowerBound(1, 1, 1);
+const cv::Scalar odk2MaskUpperBound(255, 255, 255);
```
won't these be instantiated in every compilation unit and cause multiple definition linker errors?

alexdewar replied:
It does compile ok for me. But you're right that this is wrong! It might break only if they're not inlined...

neworderofjamie commented on third_party/path.h:
```diff
@@ -350,6 +350,13 @@ inline void copy_file(const filesystem::path &from, const filesystem::path &to)
+current_path()
+{
+    char cwd[PATH_MAX];
+    return getcwd(cwd, sizeof(cwd));
```
Pretty sure this is POSIX only

alexdewar replied:
It does work on Windows (though it's technically deprecated): https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/getcwd?view=msvc-170

neworderofjamie commented on projects/snapshot_bot/snapshot_bot.cc:
```diff
@@ -474,7 +379,7 @@ class RobotFSM : FSM<State>::StateHandler
                     }
                     else {
-                        LOGW << "WARNING: Can only stream output from a perfect memory";
+                        LOGW << "Can only stream output from a perfect memory";
```
It's like it was previously just printing to ``std::cout`` or something :rofl:

neworderofjamie commented on projects/snapshot_bot/snapshot_bot.cc:
```diff
@@ -513,13 +416,13 @@ class RobotFSM : FSM<State>::StateHandler
                 if(m_Joystick.isPressed(HID::JButton::A)) {
                     // Subtract time we've already moved for from drive time
-                    m_DriveTime -= m_MoveStopwatch.elapsed();
+                    m_DriveTime -= static_cast<millisecond_t>(m_MoveStopwatch.elapsed());
```
how come this needs a cast? Doesn't the stopwatch also return a time unit?

alexdewar replied:
It doesn't, though maybe it should!

neworderofjamie commented on projects/snapshot_bot/snapshot_bot.cc:
```diff
@@ -513,13 +416,13 @@ class RobotFSM : FSM<State>::StateHandler
                 }
                 // Otherwise, if drive time has passed
-                else if(m_MoveStopwatch.elapsed() > m_DriveTime) {
+                else if(static_cast<millisecond_t>(m_MoveStopwatch.elapsed()) > m_DriveTime) {
```
Likewise

neworderofjamie replied:
If it doesn't and it's using ``std::chrono`` types, why not just stick with those - they're not used in complex units-based calculations or anything!

alexdewar replied:
Good point. I sort of hoped that it would end up making the code tidier but that didn't pan out...

alexdewar commented on projects/snapshot_bot/config.h:
```diff
@@ -20,16 +21,39 @@
+      , m_ViconTrackingPort(0)
+      , m_ViconTrackingObjectName("norbot")
+      , m_UseViconCaptureControl(false)
+      , m_ViconCaptureControlPort(0)
```
It's what clang-format does, not that that means it's a good idea! I find having all the initialisers squished onto a couple of lines a bit annoying when it comes to adding values (like I did here), but I'm not precious about it so happy to change back.

neworderofjamie replied:
fair enough - my desire to fight clang-format is probably not mega-helpful.....

alexdewar commented on projects/snapshot_bot/config.h:
```diff
@@ -170,13 +192,13 @@ class Config
-
-        cv::String testingSuffix;
-        cv::read(node["testingSuffix"], testingSuffix, m_TestingSuffix);
-        m_TestingSuffix = (std::string)testingSuffix;
```
Sorry, I should have mentioned this in the PR...

The reason for removing it was that I've changed the test data to also be in an image database format (though there may well not be actual images stored in there) and by default ``ImageDatabase``s are initialised with a timestamp so they're unique. So you'd have e.g. a folder ``training`` with your snapshots in and then your test data would be in ``training/testing/20211117_150400_001`` or whatever. (I think it makes sense to have the testing data as a subfolder in with the training data, but we could also change this.)

It would be sensible to at least the option have more explicit and readable names though. Would it be better to keep them unique (e.g. have a ``testingPrefix`` option) or just have the whole name be specified?

neworderofjamie replied:
ahhh I see - realistically, that probably makes not over-writing your old data rather less error-prone :smile: it might be nice to have an optional prefix or suffix as well though

alexdewar commented on projects/snapshot_bot/memory.cc:
```diff
@@ -29,14 +34,50 @@ MemoryBase::MemoryBase()
+
+    // Load and process images in parallel
+    // **TODO**: Add support for static mask images
+    std::atomic_int loadCount{ 0 };
```
Nope! Good spot.

neworderofjamie commented:
One thing, before this is merged is that it'd be good to check it works on the actual robot. If you can point me in the direction of the cmake PPA for Ubuntu 16, I'm happy to have a go at this.

alexdewar commented:
I've pushed the final changes that I wanted to get done on this PR (to fix #281). Other than that, I think I've addressed your comments.

Totally agree that we need to test this on a real robot. We should also probably test it for a few different configurations (e.g. with infomax and using the ODK2). There's a PPA here: https://apt.kitware.com/

I'm also happy to have a go myself. I'll probably be in tomorrow or Fri.

neworderofjamie commented:
Awesome! I'm away on Friday but will be on the office tomorrow

neworderofjamie commented:
Just trying to install the ppa but the 16.04 repository doesn't support arm64
